text/0000-natvis.md

```diff
@@ -0,0 +1,491 @@
+
+This design does not break any existing usage of cargo or rustc. This new feature
+would be strictly opt-in. The Natvis syntax may not be familiar to many Rust
+developers which may lead to a period of learning the syntax. Since this feature
+would be optional, a consumer of a crate that has natvis definitions for types
+would not need to go through this learning curve. 
+
+Not doing this would keep the existing debugging experience for external Rust crates.
+Most Rust types, outside of the standard library, do not have any debugger views
+defined for them by default which makes them difficult to interpret when viewed
+under a debugger.
+
+# Prior art
+[prior-art]: #prior-art
+
+Many debuggers and languages already address this problem. Some do so in a way
+that is more flexible than others.
+
+Briefly, we cover some of the known systems for debugger views:
+
+* Windows Debugger (WinDbg)
+* Visual Studio Debugger (VS Debugger)
+* GDB/LLDB
+
+## Windows Debugger (WinDbg)
+
+Natvis is a framework that customizes how native types appear when viewed under
+a debugger. The Visual Studio Natvis framework is supported out of the box on
+WinDbg. The debugger has the ability to load `.natvis` files via the `.nvload`
+command and directly apply them to types within loaded modules. WinDbg is also
+able to load `.natvis` files that have been embedded in the PDB for a binary and
+serve up the resulting views after applying those visualizations as well. This
+allows for a very smooth debugging experience which would not depend on any manual
+loading of Natvis files.
+
+## Visual Studio Debugger (VS Debugger)
+
+The Visual Studio Debugger also supports Natvis. Similar to WinDbg, the VS Debugger
+is also able to apply Natvis on the fly by loading user-specified `.natvis` files.
+As with WinDbg, it also supports loading `.natvis` files that were embedded in the
+PDB for a binary and automatically applying the Natvis visualizations to types from
+that binary.
+
+When using Visual Studio to build a C++ project, a developer can add a Natvis file
+via the `.vcxproj` file. To add a Natvis file to a project the following can be
+added to the `.vcxproj` file:
+
+```text
+<ItemGroup>
+  <Natvis Include="Foo.natvis" />
+</ItemGroup>
+```
+
+## GDB/LLDB
+
+GDB and LLDB also support debugger views but in a different way than WinDbg and the
+VS debugger. Natvis is not supported by either GDB or LLDB but they do support pretty
+printers. Pretty printers work in the similar manner as Natvis in which they tell
+the debugger to serve up a specific visualization when viewing a type in the debugger.
+Pretty printers are written as python scripts and then have to be imported in to the
+debugger. When a type is viewed under the debugger that has a pretty printer, that view
+is automatically shown.
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Is the `[package]` section of the `Cargo.toml` manifest the best place to add this new syntax?
```

Comment:
I think this question is out of date now

Reply:
Will remove. Thanks!